### PR TITLE
interpret: remove support for uninitialized scalars

### LINKED ISSUE #2187: Deprecate -Zmiri-allow-uninit-numbers?
I would love to deprecate the -Zmiri-allow-uninit-numbers flag some time later this year. Removing that flag would allow us to do some significant cleanup in the interpreter -- we could entirely remove the `ScalarMaybeUninit` type!

### PR DESCRIPTION
**Pull request by RalfJung**
With Miri no longer supporting `-Zmiri-allow-uninit-numbers`, we no longer need to support storing uninit data in a `Scalar`. We anyway already only use this representation for types with *initialized* `Scalar` layout (and we have to, due to partial initialization), so let's get rid of the `ScalarMaybeUninit` type entirely.

I tried to stage this into meaningful commits, but the one that changes `read_immediate` to always trigger UB on uninit is the largest chunk of the PR and I don't see how it could be subdivided.

Fixes https://github.com/rust-lang/miri/issues/2187
r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
Given https://github.com/rust-lang/rust/issues/99923, we probably want to crater this one.
@bors try

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 1]
```diff
@@ -621,13 +619,7 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
         // i.e. that we go over the `check_init` below.
```
Comment (deltragon) [inline compiler/rustc_const_eval/src/interpret/validity.rs]: Does this comment need to be updated now?

Comment (deltragon) [reply compiler/rustc_const_eval/src/interpret/validity.rs]: I saw it gets deleted in a later commit.

**Comment by RalfJung** [comment]
@craterbot check

**Comment by RalfJung** [comment]
Most of these failures are due to "Failed to get rustc version" in the libc build script... what is going on there? That can't be this PR. Out of the 10 I have checked so far, all 10 show that error. Is there some way that we don't have to check >600 regressions manually for whether they are all the same spurious failure...?

**Comment by RalfJung** [comment]
I downloaded the 'regressed' logs, removed the spurious libc things, spurious syn build failures ("trait X does not implement Y"), and some strange spurious "file not found for module", and went through the rest by hand, and they are all spurious.

**Comment by RalfJung** [comment]
Just to help diagnose the crater issue:
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-100043/retry-regressed-list.txt

**Comment by Mark-Simulacrum** [comment]
@craterbot p=1

**Comment by Dylan-DPC** [comment]
failed in [rollup ci](https://github.com/rust-lang/rust/pull/101033)

@bors r-

**Comment by RalfJung** [comment]
@Dylan-DPC so far all I can see there are Miri build failures, which are expected and allowed.

**Comment by RalfJung** [comment]
Some improvements, some regressions, does that count as them being canceled out so overall we are good?^^

The code became simpler than before so I don't quite understand why this would slow anything down.

**Comment by rylev** [comment]
Looks like [html5ever-0.26.0](https://github.com/rust-lang/rustc-perf/tree/master/collector/benchmarks/html5ever-0.26.0) has become noisy and the improvements here are making up for regressions seen in the last PR (and thus unlikely to be due to the changes here). 

The regressions in secondary benchmarks seem to be more real though. Looks like the most impacted query is `eval_to_allocation_raw`. Seems possible that that might indeed be impacted by this change (just going off the usage of `eval_to_allocation_raw` in const eval)?

Here's a run of cachegrind diff on one of the regressions:

```
 9,371,649  ???:do_rallocx
 8,716,171  ???:_rjem_je_arena_ralloc
 6,029,220  ???:_rjem_je_arena_ralloc_no_move
-3,604,458  ???:<hashbrown::set::HashSet<rustc_middle::mir::interpret::AllocId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>> as core::iter::traits::collect::Extend<rustc_middle::mir::interpret::AllocId>>::extend::<core::iter::adapters::map::Map<core::slice::iter::Iter<(rustc_target::abi::Size, rustc_middle::mir::interpret::AllocId)>, rustc_const_eval::interpret::intern::intern_shallow<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>::{closure
 3,604,458  ???:<hashbrown::map::HashMap<rustc_middle::mir::interpret::AllocId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>> as core::iter::traits::collect::Extend<(rustc_middle::mir::interpret::AllocId, ())>>::extend::<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::slice::iter::Iter<(rustc_target::abi::Size, rustc_middle::mir::interpret::AllocId)>, rustc_const_eval::interpret::intern::intern_shallow<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>::{closure
 3,276,750  ???:<rustc_const_eval::interpret::validity::ValidityVisitor<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::check_safe_pointer
 2,949,075  library/core/src/fmt/mod.rs:core::fmt::write
 2,818,180  /build/glibc-eX1tMB/glibc-2.31/string/../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:__memcpy_avx_unaligned_erms
-2,359,247  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::ref_to_mplace
```

Could it be more allocations happening in the validity visitor?

**Comment by RalfJung** [comment]
Ah! Yes there is an unconditional `format!` that is only needed on the error path. Good catch!

**Comment by rylev** [comment]
Given the results of the fix PR being positive, I'm going to mark this as triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**